Add alternative frame callback in case of Wayland client operations

arminn opened the pull request.

arminn committed:
Add alternative frame callback in case of Wayland client operations

Consider use-case when display backend operates as
a Wayland client. Guest OS as a content provider
sends buffer to display backend with an intention
to display it. Display backend should send back
page flip done event to Guest OS(i.e. that event
should provide some guarantee that buffer not only
page flipped and also scanned by display HW).
Previously, wl_callback_listener was used for that
purpose, but according to [1] it should be used for
drawing frames. I.e. those events notify when it is
time to send data for drawing, but it can't say
when drawing ends. Presentation time feedback
event [2] proposed to use instead of it.
According to documentation: A presentation_feedback
object returns an indication that a wl_surface
content update has become visible to the user.

[1] https://wayland-book.com/surfaces-in-depth/frame-callbacks.html
[2] https://cgit.freedesktop.org/wayland/wayland-protocols/tree/stable/presentation-time/presentation-time.xml#n128

Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
Correctly track Surface object

With patch c6fde3bad5 to the WlBuffer was
added logic which may call clear() method on
the Surface object. The main issue is there
are several WlBuffer objects and only one Surface.
Since raw ptr was used, memory issues may appear.
he current code adds weak_pt(not shared_pt - since
we don't want to change the overcomplicated
destruction process).

Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn commented:
Rebased on to https://github.com/xen-troops/displ_be/pull/147

al1img commented on CMakeLists.txt:
```diff
@@ -14,6 +14,7 @@ OPTION(WITH_DRM_ZCOPY "enable drm zero-copy support" ON)
 OPTION(WITH_KMS_ZCOPY "enable kms zero-copy support" ON)
 OPTION(WITH_DMABUF_ZCOPY "enable dmabuf zero-copy support" ON)
 OPTION(WITH_WAYLAND "build with wayland backend" ON)
+OPTION(WITH_WAYLAND_PRESENTATION_API "deliver callbacks in a more synchronous way using presentation time protocol" OFF)
```
Then name should be something like: `WITH_WAYLAND_PRESENTATION_FEEDBACK`?

arminn replied:
done

al1img commented on src/displayBackend/wayland/WaylandRegistry.hpp:
```diff
@@ -0,0 +1,54 @@
+
+class WaylandRegistry {
+public:
+
```
Remove empty line.

al1img commented on src/displayBackend/wayland/WaylandRegistry.hpp:
```diff
@@ -0,0 +1,54 @@
+class WaylandRegistry {
+public:
+
+    WaylandRegistry() = default;
```
Add doxygen comments for all public methods.

al1img commented on src/displayBackend/wayland/WaylandRegistry.hpp:
```diff
@@ -0,0 +1,54 @@
+
+#include <wayland-client.h>
+
+using std::map;
```
Do not use `using` in header files.

al1img commented on src/displayBackend/wayland/WaylandRegistry.cpp:
```diff
@@ -0,0 +1,59 @@
+
+void WaylandRegistry::addItem(wl_registry *registry, uint32_t id, const char *interface, uint32_t version)
+{
+    lock_guard<mutex> lock(mLock);
```
Empty line after var definition.

al1img commented on src/displayBackend/wayland/WaylandRegistry.cpp:
```diff
@@ -0,0 +1,59 @@
+void WaylandRegistry::addItem(wl_registry *registry, uint32_t id, const char *interface, uint32_t version)
+{
+    lock_guard<mutex> lock(mLock);
+    WaylandRegistryItem i = { interface, id, version, registry };
```
Put directly to emplace without intermediate var.

al1img commented on src/displayBackend/wayland/WaylandRegistry.cpp:
```diff
@@ -0,0 +1,59 @@
+{
+    lock_guard<mutex> lock(mLock);
+    auto item = mRegistry.find(interface);
+    if (item == mRegistry.end()) {
```
Empty line after var definitions.

al1img commented on src/displayBackend/wayland/WaylandRegistry.cpp:
```diff
@@ -0,0 +1,59 @@
+    if (item == mRegistry.end()) {
+        throw Exception("Invalid request to WaylandRegistry for interface " + interface, EINVAL);
+    }
+    return item->second;
```
Empty line before return.

al1img commented on src/displayBackend/wayland/Display.hpp:
```diff
@@ -171,6 +180,8 @@ class Display : public DisplayItf::Display
 
 	std::unique_ptr<XenBackend::PollFd> mPollFd;
 
+	static WaylandRegistry sWaylandRegistry;
```
Better to have it dynamic.

arminn replied:
I intentionally create it static to not have dynamic deps on the Display object.

arminn replied:
I.e. if the registry will be dynamic, you need to pass a reference to Display object to all needed objects.

al1img commented on src/displayBackend/wayland/WaylandRegistry.hpp:
```diff
@@ -0,0 +1,54 @@
+    wl_registry *registry;
+};
+
+class WaylandRegistry {
```
It looks like it is a wrapper over the map. Why not just use map member in Display class?

al1img commented on src/displayBackend/wayland/Display.hpp:
```diff
@@ -134,6 +135,14 @@ class Display : public DisplayItf::Display
+	 * Provides actual Wayland registry records
+	 * @return const reference to WaylandRegistry
+	 */
+	static const WaylandRegistry& getWaylandRegistry() {
```
You expose all unneeded functionality to other classes such as add/remote items from registry. That leads to misuse of this API. The only Display can add/remove entries. Other classes should have access only to get item method. The proper public method should be `getRegistry`.

al1img commented on CMakeLists.txt:
```diff
@@ -32,6 +33,7 @@ message(STATUS "WITH_DRM_ZCOPY                = ${WITH_DRM_ZCOPY}")
 message(STATUS "WITH_KMS_ZCOPY                = ${WITH_KMS_ZCOPY}")
 message(STATUS "WITH_DMABUF_ZCOPY             = ${WITH_DMABUF_ZCOPY}")
 message(STATUS "WITH_WAYLAND                  = ${WITH_WAYLAND}")
+message(STATUS "WITH_WAYLAND_PRESENTATION_API = ${WITH_WAYLAND_PRESENTATION_API}")
```
`WITH_WAYLAND_PRESENTATION_FEEDBACK`? Because `WITH_WAYLAND_PRESENTATION_API` too generic I guess.

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.cpp:
```diff
@@ -0,0 +1,118 @@
+
+	auto wr = Display::getWaylandRegistry().getItem(wp_presentation_interface.name);
+	mWlPresentation = (wp_presentation*)wl_registry_bind(wr.registry, wr.id, &wp_presentation_interface, 1);
+	mLog = std::string("PresentationFeedbackSurface");
```
Log initialization should be at initialization list.

arminn replied:
mLog is a part of a Surface object.... do you propose to create one more for PresentationFeedbackSurface?

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.cpp:
```diff
@@ -0,0 +1,118 @@
+PresentationFeedbackSurface::~PresentationFeedbackSurface()
+{
+	unique_lock<mutex> lock(mMutex);
+	wp_presentation_destroy(mWlPresentation);
```
Empty line after var definition.

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.cpp:
```diff
@@ -0,0 +1,118 @@
+void PresentationFeedbackSurface::framePresented(struct wp_presentation_feedback *wp_feedback, uint32_t tv_sec_hi,
+		uint32_t tv_sec_lo, uint32_t tv_nsec, uint32_t refresh_ns,
+		uint32_t seq_hi, uint32_t seq_lo, uint32_t flags) {
+	unique_lock<mutex> lock(mMutex);
```
Empty line after var definition.

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.cpp:
```diff
@@ -0,0 +1,118 @@
+	};
+
+	auto wr = Display::getWaylandRegistry().getItem(wp_presentation_interface.name);
+	mWlPresentation = (wp_presentation*)wl_registry_bind(wr.registry, wr.id, &wp_presentation_interface, 1);
```
We have many surfaces but registry should be bound only once. Better to bind/destroy it in `display` and pass wp_presentation* as parameter.

arminn replied:
Disagree with that. The proposed solution is to create a huge object(Display) with enormous dependencies ( you bind in display and need to pass it through to the correct place). The logic, in that case, will be split among several objects with all consequences regarding object destruction. The current solution accumulates all presentation logic in one object.

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.cpp:
```diff
@@ -0,0 +1,118 @@
+	DLOG(mLog, DEBUG) << "Draw";
+
+	mStoredCallback = callback;
+
```
Remove empty line.

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.cpp:
```diff
@@ -0,0 +1,118 @@
+	wl_surface_damage(mWlSurface, 0, 0,
+					  mBuffer->getWidth(),
+					  mBuffer->getHeight());
+
```
ditto

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.cpp:
```diff
@@ -0,0 +1,118 @@
+					  mBuffer->getHeight());
+
+	wl_surface_attach(mWlSurface, mBuffer->getWLBuffer(), 0, 0);
+
```
ditto

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.hpp:
```diff
@@ -0,0 +1,63 @@
+namespace Wayland {
+
+class PresentationFeedbackSurface : public Surface {
+public:
```
Why do you need two public sections?
Add oxygen comment for all public methods.

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.hpp:
```diff
@@ -0,0 +1,63 @@
+
+	static void sPresentationFeedbackHandleSyncOutput(void *data,
+		struct wp_presentation_feedback *feedback, struct wl_output *output);
+
```
Remove empty line.

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.hpp:
```diff
@@ -0,0 +1,63 @@
+		struct wp_presentation_feedback *wp_feedback, uint32_t tv_sec_hi,
+		uint32_t tv_sec_lo, uint32_t tv_nsec, uint32_t refresh_ns,
+		uint32_t seq_hi, uint32_t seq_lo, uint32_t flags);
+
```
ditto

al1img commented on src/displayBackend/wayland/PresentationFeedbackSurface.hpp:
```diff
@@ -0,0 +1,63 @@
+
+class PresentationFeedbackSurface : public Surface {
+public:
+	friend class Compositor;
```
Move friend class to private section.

arminn committed:
Destroy DisplayCommandHandler correctly

The Wayland callbacks (which eventually lead
to DisplayCommandHandler::sendFlipEvent) start
from Surface and used Connector and EventRingBuffer
objects. Prior to Connector object destruction,
all Wayland callbacks should be disabled.

Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

al1img commented on src/displayBackend/wayland/Compositor.cpp:
```diff
@@ -8,19 +8,25 @@
-Compositor::Compositor(wl_display* display, wl_registry* registry,
-					  uint32_t id, uint32_t version) :
+Compositor::Compositor(wl_display* wlDisplay, wl_registry* registry,
+					  uint32_t id, uint32_t version, Display* display) :
```
looks like Display* display is not used?

al1img commented on src/displayBackend/wayland/Display.cpp:
```diff
@@ -437,9 +437,22 @@ void Display::registryHandler(wl_registry *registry, uint32_t id,
+#ifdef WITH_WAYLAND_PRESENTATION_FEEDBACK
+	if (interface == wp_presentation_interface.name) {
+		mWlPresentation = (wp_presentation*)wl_registry_bind(registry, id, &wp_presentation_interface, 1);
+		if(mCompositor)
```
Space after if, use brackets for one line condition also.

al1img commented on src/displayBackend/wayland/Display.cpp:
```diff
@@ -437,9 +437,22 @@ void Display::registryHandler(wl_registry *registry, uint32_t id,
-		mCompositor.reset(new Compositor(mWlDisplay, registry, id, version));
+		mCompositor.reset(new Compositor(mWlDisplay, registry, id, version, this));
+#ifdef WITH_WAYLAND_PRESENTATION_FEEDBACK
+		if(mWlPresentation) {
```
Space after if

al1img commented on src/displayBackend/wayland/Compositor.cpp:
```diff
@@ -46,8 +52,11 @@ Compositor::~Compositor()
 	LOG(mLog, DEBUG) << "Create surface";
-
+#ifdef WITH_WAYLAND_PRESENTATION_FEEDBACK
+	return SurfacePtr(new PresentationFeedbackSurface(mWlCompositor, mWlPresentation));
```
What if setPresentation is not called (weston doesn't send presentation registry for example)?
